Feature/claudia 4723

chemaper opened the pull request:
#### Reviewers

@flopezag @jframos 
#### Description

As part of CLAUDIA-4723, the script to rename images has been updated to mark images as obsolete.

However, adding an operation with a new tool is against CLAUDIA-4723 (unified process) and therefore the code has been integrated also in glancesync. It is managed as an state, not as an operation. It is enough with rename the images to the '_obsolete' suffix and make then private in the master region.  

Documentation and test are included.
#### Testing

Locally

chemaper committed:
update example script to make possible deprecate an image

chemaper committed:
new:user:integrate support for obsolete images

jframos commented on tests/resources/obsolete.result/backup_Valladolid.csv:
```diff
@@ -0,0 +1,7 @@
+Valladolid,image03_obsolete,004,active,1000,checksum03,tenant1id,False,"{'prop1':'key1', 'prop2':'key2'}"
+Valladolid,image04_obsolete,005,active,1000,checksum04,tenant1id,False,"{'prop1':'key1', 'prop2':'key2'}"
+Valladolid,image05_obsolete,006,active,1000,checksum05,tenant1id,False,"{'prop1':'key1', 'prop2':'key2'}"
+Valladolid,image06_obsolete,007,active,1000,checksum06,tenant1id,False,"{'prop1':'key1', 'prop2':'key2'}"
```
New line at the end of this file.

jframos commented on glancesync/glancesync_config.py:
```diff
@@ -187,6 +187,9 @@ def __init__(self, configuration_path=None, stream=None):
                     target['only_tenant_images'] = True
 
+                target['obsolete_syncprops'] = configparser.getset(
+                        section, 'obsolete_syncprops')
```
Add the new property **obsolete_syncprops** and its documentation to the template config file `conf/glancesync.conf`.
Perhaps it is interesting to have a boolean property to sync or not _obsolete_ images...

chemaper replied:
The boolean property was added and the documentation updated.

I've removed the file conf/glancesync.conf and add a script to generate the config file from the README.rst.

jframos commented:
As result of this new feature, we should create some new acceptance tests.

flopezag commented on README.rst:
```diff
@@ -185,6 +197,35 @@ are copied. Be aware that system property *is_public* must not be included in
+region to synchronise that is already renamed but without the other changes
+propagated.
+
+Actually, there are some checks to do before propagating the changes of an
```
Actually or Currently?

chemaper replied:
My idea in Spanish was "realmente". But I'll remove "actually," I have checked that the sense I had though is not quite correct.

flopezag commented on README.rst:
```diff
@@ -185,6 +197,35 @@ are copied. Be aware that system property *is_public* must not be included in
+Actually, there are some checks to do before propagating the changes of an
+obsoleted image:
+
+* Are the two images the same? The checksum is compared and only if they are
```
The checksums are

flopezag commented on README.rst:
```diff
@@ -185,6 +197,35 @@ are copied. Be aware that system property *is_public* must not be included in
+obsoleted image:
+
+* Are the two images the same? The checksum is compared and only if they are
+  equals the change is done.
```
the same the change is done.

chemaper committed:
fix English errata @minor

chemaper committed:
Merge branch 'develop' of https://github.com/telefonicaid/fiware-glancesync into feature/CLAUDIA-4723

Conflicts:
	README.rst

flopezag commented on examples/renameimage.py:
```diff
@@ -28,6 +28,65 @@
 
+"""This code renames an image in all the regions. The image must exist in
+the master region (it is OK if the image has been already renamed). If the
+image does not exists in the remote regions, it is silently ignored.
```
does not exist

chemaper replied:
changed.

chemaper committed:
add configuration parameter to control if obsolete images are processed @minor

chemaper committed:
add comments to the configuration file with the new options

chemaper committed:
new:documentation: remove example configuration file, add a script to extract it from README.rst.
 This way the possibility that the two files are out of sync is avoided.

chemaper committed:
add new line @minor

chemaper committed:
add missing resources files @minor

flopezag commented on examples/renameimage.py:
```diff
@@ -28,6 +28,65 @@
+way to deprecated an image: adding the suffix "deprecated" and changing
+is_public to False.
+
+If there is more than image in a region matching the name and checksum, an
```
more than one image

flopezag commented on examples/renameimage.py:
```diff
@@ -28,6 +28,65 @@
+    :param image_master:  the master image
+    :return: True if the image needs an update, False otherwise.
+    """
+    if image.is_public != image_master.is_public:
```
I do not like to have more than one return in the same method, try to rewrite the code.

chemaper replied:
Ok, the function has been rewritten.

flopezag commented on examples/renameimage.py:
```diff
@@ -38,44 +97,100 @@
+                msg = 'There are more than an image with the new name in the' \
+                      ' master region'
+                logging.error(msg)
+                sys.exit(-1)
```
why sys.exit(-1) and not raise an Exception?

chemaper replied:
Because it is a script not a module. Scripts does not raise exceptions, print a message and exit with an error code. Exceptions are for developers not for users, in general a program that shows an exception (not in debug mode) instead of a clean error message and an error code, IMHO make a poor impression, of a not finish work.

flopezag replied:
But you are not printing you are logging a message

chemaper replied:
logging.error by default uses the root handler, that prints a message to sys.stderr.

flopezag replied:
NTC

flopezag commented on examples/renameimage.py:
```diff
@@ -38,44 +97,100 @@
+                msg = 'There are more than an image with the old name in the' \
+                      ' master region'
+                logging.error(msg)
+                sys.exit(-1)
```
why sys.exit and not raise an exception?

flopezag commented on examples/renameimage.py:
```diff
@@ -38,44 +97,100 @@
+                new_image = image
+
+    if old_image and new_image:
+        msg = 'Both an image with the new name and the old new exists in the' \
```
'Both an image with the new name and the old new exists in the master region'
->
'There is an image with the same old and new name in the master region'

flopezag commented on examples/renameimage.py:
```diff
@@ -38,44 +97,100 @@
+        master_image = new_image
+    elif old_image:
+        master_image = old_image
+        old_image.name = sys.argv[2]
```
If I understand the algorithm, in this point you are taking the old_image.name the name of the second parameter that previously you are comparing for the new_image, it is something confuse.

chemaper replied:
The confusion is probable because a bad chosen variable names. old_image should be image_with_old_name, new_image, image_with_new_name.

The idea is: search for images with old/new name. It both exist, this is an error. If only image with the new name exits, the image is already renamed, use this image as reference to use the checksum, is_property and other properties. If only image with the old name exists, the same, but first update the name (rename it)

I will change the names

flopezag commented on examples/renameimage.py:
```diff
@@ -38,44 +97,100 @@
+        print 'Renamed image in master region'
+    else:
+        msg = 'Neither an image with the old name nor a one with the new name'\
+            ' exists in the master region'
```
exist

chemaper replied:
I'll change it, although according the OALD examples, both forms are correct. This is an example in the neither entry:
  Neither the TV nor the video actually work/works

flopezag commented on examples/renameimage.py:
```diff
@@ -38,44 +97,100 @@
+        msg = 'Neither an image with the old name nor a one with the new name'\
+            ' exists in the master region'
+        logging.error(msg)
+        sys.exit(-1)
```
Why not raise exception?

flopezag commented on examples/renameimage.py:
```diff
@@ -38,44 +97,100 @@
+                if image_already_renamed:
+                    if metadata_outdated(image_already_renamed, master_image):
+                        update_image(image_already_renamed, master_image)
+                        print 'medatadata updated in already renamed image'
```
This sentence will produce a technical debt, please avoid usage of print statement and change to:

print('medatadata updated in already renamed image')

flopezag commented on examples/renameimage.py:
```diff
@@ -38,44 +97,100 @@
+                        update_image(image_already_renamed, master_image)
+                        print 'medatadata updated in already renamed image'
+                    else:
+                        print 'Already renamed'
```
idem

flopezag commented on examples/renameimage.py:
```diff
@@ -38,44 +97,100 @@
+                    else:
+                        print 'Already renamed'
+                else:
+                    print 'Not found.'
```
idem

flopezag commented on glancesync/glancesync_image.py:
```diff
@@ -251,6 +252,9 @@ def is_synchronisable(
         :param metadata_condition: expression to evaluate if the image is sync.
         :return:
```
Provide only one exit point from the function

chemaper replied:
done

flopezag commented on glancesync/glancesync_region.py:
```diff
@@ -241,3 +241,107 @@ def image_list_to_sync(self, images_master_region, images_region):
+
+    def image_list_to_obsolete(self, images_master_region, images_region,
+                                obsolete_syncprops=None):
+        """Obtain a list of images in the region that must be mark as obsolete,
```
must be marked...

jframos commented:
Should we remove the directory `conf`?

flopezag commented on glancesync/glancesync_region.py:
```diff
@@ -241,3 +241,107 @@ def image_list_to_sync(self, images_master_region, images_region):
+    def image_list_to_obsolete(self, images_master_region, images_region,
+                                obsolete_syncprops=None):
+        """Obtain a list of images in the region that must be mark as obsolete,
+        i.e. the suffix must be renamed to '_obsolete' and make it private.
```
to '_obsolete' and made

chemaper committed:
rewrite function to avoid multiple returns @minor

chemaper committed:
rename variables to make code more clear @minor

chemaper committed:
fix English errata @minor

chemaper committed:
fix English errata @minor

chemaper committed:
function recoded to have an only return @minor

chemaper committed:
changed print str -> print(str) and print str, to sys.stdout.write(str) sys.stdout.flush()

flopezag commented on examples/renameimage.py:
```diff
@@ -36,48 +104,109 @@
 
     for region in regions:
-        print region + ' ',
+        sys.stdout.write(region + ' ')
```
Why do not you take the message for each for step an the last sentence of the loop print it? I mean:

```
msg = region + ' '
...
msg = msg + 'Not renamed.' 

....

print(msg)
```

chemaper replied:
Because the region name is printed before making the call. The call may last a while or even fail. This is similar to the init start scripts, where a service name is printed before starting and then OK or FAIL

flopezag replied:
NTC

flopezag commented on scripts/generate_config_file.py:
```diff
@@ -0,0 +1,41 @@
+
+line = lines_iter.next()
+while line != '.. glancesync_conf_end':
+    print line[1:]
```
do not forget this print...

chemaper committed:
fix print @minor

chemaper committed:
move obsolete images treatment before getting data @minor
  this is only to make easy both apply before obsolete and then sync altough now this is not done.

flopezag commented:
LGTM

chemaper committed:
remove empty conf folder @minor

chemaper commented:
deleted empty conf folder.

jframos commented:
LGTM
